languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -0,0 +1,477 @@
+# Roles 2
+from typing import Any, List, Set
+from dataclasses import dataclass
+
+from oso import Variable
+
+from sqlalchemy.types import Integer, String
+from sqlalchemy.schema import Column, ForeignKey
+from sqlalchemy import inspect
+
+
+# Starting with something like this, tbd
+
+
+@dataclass
+class Relationship:
+    child_python_class: Any
+    child_type: str
+    child_table: str
+    parent_python_class: Any
+    parent_type: str
+    parent_table: str
+    parent_selector: Any
+    parent_field: str
+
+
+@dataclass
+class Permission:
+    python_class: Any
+    name: str
+
+
+@dataclass
+class Role:
+    name: str
+    python_class: Any
+    permissions: List[Permission]
+    implied_roles: List[str]
+
+
+@dataclass
+class Resource:
+    name: str
+    python_class: Any
+    actions: Set[str]
+    roles: Set[str]
+
+
+def ensure_configured(func):
+    def wrapper(self, *args, **kwargs):
+        if not self.configured:
+            self._configure()
+        func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class OsoRoles:
+    def __init__(self, oso, sqlalchemy_base, user_model):
+        # @Q: Is this where I should create the models?
+        # Would this even work? Do these have to happen at a
+        # certain time or something to get created in the database?
+
+        self.session = None
+
+        user_pk_type = inspect(user_model).primary_key[0].type
+        user_pk_name = inspect(user_model).primary_key[0].name
+        user_table_name = user_model.__tablename__
+
+        # Tables for the management api to save data.
+        class UserRole(sqlalchemy_base):
+            __tablename__ = "user_roles"
+            id = Column(Integer, primary_key=True)
+            user_id = Column(
+                user_pk_type, ForeignKey(f"{user_table_name}.{user_pk_name}")
+            )
+            resource_type = Column(String)
+            resource_id = Column(String)  # Most things can turn into a string lol.
+            role = Column(String)
+
+        class Permission(sqlalchemy_base):
+            __tablename__ = "permissions"
+            id = Column(Integer, primary_key=True)
```

Comment:
Any reason permission has an id col but not role?

Reply:
role just uses it's name as the id since it's globally unique, permission has no such builtin key